api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto

```diff
@@ -118,19 +118,27 @@ message ExternalProcessor {
   // for the list of supported attributes and their types.
   repeated string response_attributes = 6;
 
-  // [#not-implemented-hide:]
   // Specifies the timeout for each individual message sent on the stream and
   // when the filter is running in synchronous mode. Whenever
   // the proxy sends a message on the stream that requires a response, it will
   // reset this timer, and will stop processing and return an error (subject
-  // to the processing mode) if the timer expires. There is no timeout when
-  // the filter is running in asynchronous mode. Default is 200 ms.
+  // to the processing mode) if the timer expires before a matching response.
+  // is received. There is no timeout when the filter is running in asynchronous
+  // mode. Default is 200 milliseconds.
   google.protobuf.Duration message_timeout = 7;
 
   // [#not-implemented-hide:]
   // Optional additional prefix to use when emitting statistics. This allows to distinguish
   // emitted statistics between configured *ext_proc* filters in an HTTP filter chain.
   string stat_prefix = 8;
+
+  // Specifies the timeout for the overall gRPC conversation between the proxy
+  // and the external processor. This timer is set when the conversation is
+  // started with the external processor and cleared when either the proxy or
+  // the external processor closes the stream. If it expires before that, then
+  // the proxy will return an error to the downstream client. Default is
+  // 10 seconds.
+  google.protobuf.Duration conversation_timeout = 9;
```

Comment:
@yuval-k this reminds me of the ext_authz timeout snafu. Any advice for @gbrail on implementing a similar timeout in a different filter that uses gRPC calls and wants to express stream timeouts?

Reply:
conversation_timeout seems to limit a single interaction and message_timeout limits the whole processing? I assume the idea is to protect against slow processing server?

If the request is unbuffered by a previous filter, a slow downstream might trigger the message_timeout unintentionally?

Reply:
Yes, but it's the other way around -- message_timeout limits a single request/response exchange on the bidi gRPC stream between the proxy and the external processing server. conversation_timeout starts when the stream is started (usually in decodeHeaders) and ends when the stream is closed, which could be as late as encodeTrailers.

Given the way that gRPC seems to work in Envoy, the second timeout -- conversation_timeout -- is the same thing as the overall gRPC timeout that the gRPC client supports.

Yes, the idea is to prevent the proxy from being ground to a halt by a slow processing server. The message_timeout is doubly important because it gives the proxy a safety valve in case the processing server fails to respond to a request as dictated by the protocol because it crashed or because the programmer neglected to code it correctly and send back a response.

Reply:
I answered this via email which apparently put the answer at the bottom of the comments stream and not here -- now I know better.

See my last comment, then, but basically, the timeout could be triggered by a slow downstream OR a slow upstream.

Does Envoy have a built-in timeout for an HTTP request/response? If so then the default for this timeout should IMO be higher than that.

Otherwise, does it make sense therefore for this timeout to be eliminated? That would mean that when we create the gRPC stream in client_impl.cc using Grpc::AsyncClient that we would not set any timeout at all, but would rely on the filter being eventually destroyed if there is some other timeout.

Reply:
i believe envoy does has timeouts for requests and responses i think default request timeout is 15s.

As `message_timeout` has a default, I'm not sure if `conversation_timeout` is needed regardless of other envoy timeouts? What scenario would conversation timeout protected that message timeout wouldn't?

Reply:
This seemed terribly important but as I think about it I can't think of a reason for an additional timeout either, since we have a per-message timeout on stream messages. So let me take it out since the message timeout is going to give us what we need anyway.

Reply:
I took out the conversation timeout, which leaves this PR to just implement the message timeout properly.